aggregator/src/cache.rs

```diff
@@ -206,3 +207,240 @@ impl PeerAggregatorCache {
             .find(|peer| peer.endpoint() == endpoint && peer.role() == role)
     }
 }
+
+#[derive(Debug)]
+pub struct TaskAggregatorCache<C: Clock> {
+    datastore: Arc<Datastore<C>>,
+    report_writer: Arc<ReportWriteBatcher<C>>,
+    cache: Cache<TaskId, TaskAggregatorCacheEntry<C>>,
+    cache_none: bool,
+}
+
+/// A cache entry. None indicates that the entry has not been initialized yet.
+type TaskAggregatorCacheEntry<C> = Arc<Mutex<Option<TaskAggregatorRef<C>>>>;
+
+/// An Arc reference to a TaskAggregator. None indicates that there is no such task aggregator in
+/// the database.
+type TaskAggregatorRef<C> = Option<Arc<TaskAggregator<C>>>;
+
+pub const TASK_AGGREGATOR_CACHE_DEFAULT_TTL: Duration = Duration::from_secs(600);
+pub const TASK_AGGREGATOR_CACHE_DEFAULT_CAPACITY: u64 = 10_000;
+
+impl<C: Clock> TaskAggregatorCache<C> {
+    pub fn new(
+        datastore: Arc<Datastore<C>>,
+        report_writer: ReportWriteBatcher<C>,
+        cache_none: bool,
+        capacity: u64,
+        ttl: Duration,
+    ) -> Self {
+        Self {
+            datastore,
+            report_writer: Arc::new(report_writer),
+            cache: CacheBuilder::new(capacity).time_to_live(ttl).build(),
+            cache_none,
+        }
+    }
+
+    pub async fn get(&self, task_id: &TaskId) -> Result<TaskAggregatorRef<C>, Error> {
+        // Step one: grab the existing entry for this task, if one exists. If there is no existing
+        // entry, write a new uninitialized entry.
+        let entry = self
+            .cache
+            .get_with(*task_id, async { Default::default() })
+            .await;
+
+        // Step two: if the entry is uninitialized or expired, fill it via a database query. Concurrent
+        // callers requesting the same task will contend over this lock while awaiting the result of
+        // the DB query, ensuring that in the common case only a single query will be made for each
+        // task.
+        let task_aggregator = {
+            let mut entry = entry.lock().await;
+
+            if entry.is_none() {
+                *entry = Some(
+                    self.datastore
+                        .run_tx("task_aggregator_get_task", |tx| {
+                            let task_id = *task_id;
+                            Box::pin(async move { tx.get_aggregator_task(&task_id).await })
+                        })
+                        .await?
+                        .map(|task| TaskAggregator::new(task, Arc::clone(&self.report_writer)))
+                        .transpose()?
+                        .map(Arc::new),
+                );
+            }
+
+            // Unwrap safety: we've ensured that the entry is initialized in the previous if statement.
+            let task_aggregator = entry.as_ref().unwrap();
+            if !self.cache_none && task_aggregator.is_none() {
+                self.cache.invalidate(task_id).await;
+            }
+            task_aggregator.clone()
```

Comment:
I noted https://github.com/divviup/janus/pull/3029#discussion_r1576775414, but it occurs to me we can make this an `Arc<Option<_>>` instead of an `Option<Arc<_>>` to achieve this clarity.

Reply:
ah, apologies for missing your comment! another option would be to do something like: `task_aggregator.as_ref().map(Arc::clone)` (though that's a bit of a mouthful)